src/neuroconv/tools/spikeinterface/spikeinterface.py

```diff
@@ -1436,3 +1459,157 @@ def write_sorting(
+    units_description : str (optional)
+    """
+    if metadata is None:
+        metadata = dict()
+    assert (
+        nwbfile_path is None or nwbfile is None
+    ), "Either pass a nwbfile_path location, or nwbfile object, but not both!"
+    if nwbfile is not None:
+        assert isinstance(nwbfile, pynwb.NWBFile), "'nwbfile' should be a pynwb.NWBFile object!"
+
+    # retrieve templates and stds
+    template_means = waveform_extractor.get_all_templates()
+    template_stds = waveform_extractor.get_all_templates(mode="std")
+    sorting = waveform_extractor.sorting
+    if unit_ids is not None:
+        unit_indices = sorting.ids_to_indices(unit_ids)
+        template_means = template_means[unit_indices]
+        template_stds = template_stds[unit_indices]
+
+    # metrics properties (quality, template) are temporarily added as properties
+    # to be written in the Units table
+    tmp_properties = []
+    if waveform_extractor.is_extension("quality_metrics"):
+        qm = waveform_extractor.load_extension("quality_metrics").get_data()
+        for prop in qm.columns:
+            if prop not in sorting.get_property_keys():
+                sorting.set_property(prop, qm[prop])
+                tmp_properties.append(prop)
+    if waveform_extractor.is_extension("template_metrics"):
+        tm = waveform_extractor.load_extension("template_metrics").get_data()
+        for prop in tm.columns:
+            if prop not in sorting.get_property_keys():
+                sorting.set_property(prop, tm[prop])
+                tmp_properties.append(prop)
+
+    assert write_as in [
+        "units",
+        "processing",
+    ], f"Argument write_as ({write_as}) should be one of 'units' or 'processing'!"
+    if write_as == "units":
+        assert units_name == "units", "When writing to the nwbfile.units table, the name of the table must be 'units'!"
+    write_in_processing_module = False if write_as == "units" else True
+
+    with make_or_load_nwbfile(
+        nwbfile_path=nwbfile_path, nwbfile=nwbfile, metadata=metadata, overwrite=overwrite, verbose=verbose
+    ) as nwbfile_out:
+        if write_electrical_series:
+            if waveform_extractor._recording is None:
+                assert recording is not None, (
+                    "To write electrical series, the waveform_extractor needs to have a recording attached. "
+                    "For recordingless mode, you can use the 'recording' argument."
+                )
+            write_recording(waveform_extractor._recording, nwbfile_out, metadata, **write_electrical_series_kwargs)
+        else:
+            # we need add_electrodes_from_waveforms
+            if waveform_extractor._recording is None:
+                if recording is not None:
+                    assert isinstance(recording, BaseRecording), "'recording' needs to be a BaseRecording object."
+            else:
+                recording = waveform_extractor.recording
```

Comment:
OK so this is the case where the user does not want to write the electrical series, but we still require a recording to be attached to the `WaveformExtractor` so we can use the electrode info?

Is it possible to have a `WaveformExtractor` that does not have either the `_recording` or `recording` attribute, or is at least one always `None`?

Are the two cases to support a previous version of SI or what else causes the distinction between the private and non-private version?

Reply:
Dev branch CI: https://github.com/catalystneuro/neuroconv/actions/runs/3631606377/jobs/6126460171

is fine with everything